text/0000-template.md

```diff
@@ -0,0 +1,341 @@
+
+We will rarely have a `not-validated` block/transaction because they are basically verified as soon as they are received. No transaction/block saved in the storage might be `not-validated`.
+
+## Optimizations & Sources of blocks and transactions
+
+The fullnode receives new blocks and transactions from two different sources: (i) the Sync Protocol, and (ii) its APIs.
+
+When the source is the Phase 1 of the Sync Protocol, we may optimize and only verify the PoW. This optimization assumes that nobody can generate another transaction with the same hash as the original transaction (because it would imply finding a sha256 conflict). The fullnode currently does not use this optimization and verifies every block/transaction received.
+
+When a transaction is received by the API, it must be fully validated. If it is not possible, the fullnode must reject it.
+
+## Measure of Sync Progress
+
+Each peer connected to the fullnode has its own progress indicator. As we know the height of the best block of a given peer, we know how many blocks we still need to download to finish the sync. We say that a block is fully downloaded when the block and all its transactions have been sucessfully downloaded. We can say that `sync_percentage = #(remaining blocks) / #(total blocks) = 1 - downloaded_blocks / height_best_block`.
+
+We say that we are synced to a peer when all blocks between the genesis and the best block have been fully downloaded. The mempool does not be affect the sync status.
+
+TODO Global sync progress? We may get the score the best block, so we can calculate how many blocks we need to download until we reach it (regardless of the peers).
+
+TODO Measure progress of the mempool?
+
+TODO Measure progress of the block? We may include the number of transactions inside a block as a metadata.
+
+## How does a reorg propagates through the network?
+
+The bestchain is the chain with the highest score among all chains. Notice that a new bestchain does not affect Phase 1 because it must have all checkpoints inside it. So, during Phase 2, our fullnode will verify that the peer has a different best block.
+
+In this case, during the first sub-step, it will find out the most recent block we have in common with the peer. Then, in the second sub-step, the fullnode will download the blocks of the new bestchain.
+
+After the last block is fully downloaded, the consensus algorithm will confirm that this chain has a higher score and will set it as the new bestchain.
+
+As soon as the new bestchain is set, all other peers connected to us will do the same during their Phase 2 with us. This way, the new bestchain will be propagated through the whole network.
+
+## Double spending problem
+
+The double spending problem is solved by the following rules:
+
+1. There cannot be any transactions spending the same output in the bestchain. If it happens, the blocks will be invalidated and they cannot be part of the bestchain.
```

Comment:
What do you mean about "the blocks will be invalidated"? The txs will be invalidated, right?

Reply:
I meant "voided". Fixed! Thanks!